scripts/split_repo_migration/git-migrate-history.sh

```diff
@@ -0,0 +1,156 @@
+  BRANCH=$(basename ${SOURCE_REPO})-migration
+fi
+export FILTER_BRANCH_SQUELCH_WARNING=1
+
+# create a working directory
+WORKDIR="$(mktemp -d -t code-migration-XXXXXXXXXX)"
+echo "Created working directory: ${WORKDIR}"
+
+pushd "${WORKDIR}"  # cd into workdir
+
+echo "Cloning source repository: ${SOURCE_REPO}"
+git clone "git@github.com:${SOURCE_REPO}.git" source-repo
+
+pushd source-repo
+git remote remove origin
+
+# prune only files within the specified directory
+if [[ ! -z "${SOURCE_PATH}" ]]
+then
+  echo "Pruning commits only including path: ${SOURCE_PATH}"
+  git filter-branch \  # rewrites history... unsafe
+    --prune-empty \
+    --subdirectory-filter "${SOURCE_PATH}"
+fi
+
+if [[ ! -z "${IGNORE_FOLDERS}" ]]
+then
+  echo "Ignoring folder: ${IGNORE_FOLDERS}"
+  mkdir -p "${WORKDIR}/filtered-source"
+  FOLDERS=$(echo ${IGNORE_FOLDERS} | tr "," " ")
+  # remove files/folders we don't want
+  FILTER="(rm -rf ${FOLDERS} || true)"
+  if [[ ! -z "${KEEP_FILES}" ]]
+  then
+    KEEP_FILES_SPACES=($(echo ${KEEP_FILES} | tr "," " "))
+    LAST_ELEMENT=$(( ${#KEEP_FILES_SPACES[@]} - 1 ))
+    KEEP_FILE_COMMANDS=""
+    for file in "${KEEP_FILES_SPACES[@]}"
+    do
+      if [[ $file == "${KEEP_FILES_SPACES[$LAST_ELEMENT]}" ]]
+      then
+        KEEP_FILE_COMMANDS+="git checkout -- ${file} 2>/dev/null || true"
+      else 
+        KEEP_FILE_COMMANDS+="git checkout -- ${file} 2>/dev/null || true; "
+      fi   
+    done
+    # restore files to keep, silence errors if the file doesn't exist
+    FILTER="${FILTER}; ${KEEP_FILE_COMMANDS}"
+  fi
+  git filter-branch \
+    --force \
+    --prune-empty \
+    --tree-filter "${FILTER}"
+fi
+
+# reorganize the filtered code into the desired target locations
+if [[ ! -z "${TARGET_PATH}" ]]
+then
+  echo "Moving files to destination path: ${TARGET_PATH}"
+  git filter-branch \
+    --force \
+    --prune-empty \
+    --tree-filter \
+      "shopt -s dotglob; mkdir -p ${WORKDIR}/migrated-source; mv * ${WORKDIR}/migrated-source; mkdir -p ${TARGET_PATH}; mv ${WORKDIR}/migrated-source/* ${TARGET_PATH}"
+fi
+
+# back to workdir
+popd
+
+# merge histories
+echo "Cloning target repository: ${SOURCE_REPO}"
+git clone "git@github.com:${TARGET_REPO}.git" target-repo
+pushd target-repo
+
+git remote add --fetch migration ../source-repo
+git checkout -b "${BRANCH}"
+git merge --allow-unrelated-histories migration/main --no-edit
+
+echo "Success"
+
+popd # back to workdir
+
+# Do a diff between source code split repo and migrated code.
+git clone "git@github.com:${SOURCE_REPO}.git" source-repo-validation  # Not ideal to clone again.
+rm -rf source-repo-validation/.git  # That folder is not needed for validation.
+
+if diff -r target-repo/"${TARGET_PATH}" source-repo-validation; then
+  echo "No diff"
+else
+  diff -r target-repo/"${TARGET_PATH}" source-repo-validation > "diff.txt"
+  echo "Diff non-empty. See ${WORKDIR}/diff.txt"
+  $EXIT 1
+fi
```

Comment:
```suggestion
DIFF_FILE="${WORKDIR}/diff.txt"
if diff -r "target-repo/${TARGET_PATH}" source-repo-validation > "${DIFF_FILE}" ; then
  echo "No diff"
else
  echo "Diff non-empty. See ${DIFF_FILE}"
  $EXIT 1
fi
```

I think the above  should work; the redirection is separate from the exit code.

You could even shorten this as follows if you want:

```suggestion
DIFF_FILE="${WORKDIR}/diff.txt"
[[ diff -r "target-repo/${TARGET_PATH}" source-repo-validation > "${DIFF_FILE}" ]]  \
  &&   echo "No diff" \
  || { echo "Diff non-empty. See ${DIFF_FILE}" ;  $EXIT 1 ; }

```

Reply:
Thanks for suggestion. Added.